File: master-node-docker/sentinel/helpers/eth.py
# coding=utf-8
from hashlib import md5

from .referral import add_bonus
from .referral import get_vpn_sessions
from ..config import COINBASE_ADDRESS
from ..config import COINBASE_PRIVATE_KEY
from ..config import LIMIT_100MB
from ..config import LIMIT_10MB
from ..config import REFERRAL_DUMMY
from ..config import SESSIONS_SALT
from ..db import db
from ..eth import erc20_manger
from ..eth import eth_manager
from ..eth import vpn_service_manager


def get_encoded_session_id(account_addr, index):
    account_addr = str(account_addr).lower().encode()
    index = str(index).encode('utf-8')
    session_id = md5(account_addr + index + SESSIONS_SALT).hexdigest()

    return session_id


class ETHHelper(object):
    def create_account(self, password):
        error, account_addr, private_key, keystore = eth_manager['main'].create_account(password)

        return error, account_addr, private_key, keystore

    def get_account_addr(self, private_key):
        error, account_addr = eth_manager['main'].get_address(private_key)
        account_addr = account_addr[2:]

        return error, account_addr

    def get_tx_receipt(self, tx_hash, net):
        error, receipt = None, None
        if net == 'main':
            error, receipt = eth_manager['main'].get_tx_receipt(tx_hash)
        elif net == 'rinkeby':
            error, receipt = eth_manager['rinkeby'].get_tx_receipt(tx_hash)

        return error, receipt

    def get_tx(self, tx_hash, net):
        error, tx = None, None
        if net == 'main':
            error, tx = eth_manager['main'].get_tx(tx_hash)
        elif net == 'rinkeby':
            error, tx = eth_manager['rinkeby'].get_tx(tx_hash)

        return error, tx

    def get_tx_count(self, account_addr, net):
        tx_count = 0
        if net == 'main':
            _, tx_count = eth_manager['main'].get_transaction_count(
                account_addr)
        elif net == 'rinkeby':
            _, tx_count = eth_manager['rinkeby'].get_transaction_count(
                account_addr)

        return tx_count

    def get_balances(self, account_addr):
        balances = {
            'main': {
                'eths': None,
                'sents': None
            },
            'rinkeby': {
                'eths': None,
                'sents': None
            }
        }
        _, balances['main']['eths'] = eth_manager['main'].get_balance(account_addr)
        _, balances['main']['sents'] = erc20_manger['main']['SENT'].get_balance(account_addr)
        _, balances['rinkeby']['eths'] = eth_manager['rinkeby'].get_balance(account_addr)
        _, balances['rinkeby']['sents'] = erc20_manger['rinkeby']['SENT'].get_balance(account_addr)

        return balances

    def transfer(self, from_addr, to_addr, amount, symbol, private_key, net):
        if symbol == 'ETH':
            error, tx_hash = self.transfer_eths(from_addr, to_addr, amount, private_key, net)
        else:
            error, tx_hash = self.transfer_erc20(from_addr, to_addr, amount, symbol, private_key, net)

        return error, tx_hash

    def transfer_eths(self, from_addr, to_addr, amount, private_key, net):
        error, tx_hash = None, None
        if net == 'main':
            error, tx_hash = eth_manager['main'].transfer_amount(from_addr, to_addr, amount, private_key)
        elif net == 'rinkeby':
            error, tx_hash = eth_manager['rinkeby'].transfer_amount(from_addr, to_addr, amount, private_key)

        return error, tx_hash

    def transfer_erc20(self, from_addr, to_addr, amount, symbol, private_key, net):
        error, tx_hash = erc20_manger[net][symbol].transfer_amount(from_addr, to_addr, amount, private_key)

        return error, tx_hash

    def raw_transaction(self, tx_data, net):
        error, tx_hash = None, None
        if net == 'main':
            error, tx_hash = eth_manager['main'].send_raw_transaction(tx_data)
        elif net == 'rinkeby':
            error, tx_hash = eth_manager['rinkeby'].send_raw_transaction(tx_data)

        return error, tx_hash

    def get_initial_payment(self, account_addr):
        error, is_paid = vpn_service_manager.get_initial_payment(account_addr)

        return error, is_paid

    def get_due_amount(self, account_addr):
        error, due_amount = vpn_service_manager.get_due_amount(account_addr)

        return error, due_amount

    def get_vpn_sessions_count(self, account_addr):
        error, sessions_count = vpn_service_manager.get_vpn_sessions_count(account_addr)

        return error, sessions_count

    def get_latest_vpn_usage(self, account_addr):
        error, usage = None, None
        error, sessions_count = self.get_vpn_sessions_count(account_addr)
        if (error is None) and (sessions_count > 0):
            session_id = get_encoded_session_id(account_addr, sessions_count - 1)
            error, _usage = vpn_service_manager.get_vpn_usage(account_addr, session_id)
            if error is None:
                usage = {
                    'id': session_id,
                    'account_addr': str(_usage[0]).lower(),
                    'received_bytes': _usage[1],
                    'session_duration': _usage[2],
                    'amount': _usage[3],
                    'timestamp': _usage[4],
                    'is_paid': _usage[5]
                }

        return error, usage

    def get_vpn_usage(self, account_addr, device_id=None):
        error, usage = None, {
            'due': 0,
            'stats': {
                'received_bytes': 0,
                'duration': 0,
                'amount': 0
            },
            'sessions': []
        }
        if device_id:
            sessions = get_vpn_sessions(device_id)
            sessions_count = len(sessions)
            for index in range(0, sessions_count):
                session = sessions[index]
                usage['stats']['received_bytes'] += session['sent_bytes']
                usage['stats']['duration'] += session['session_duration']
                usage['stats']['amount'] += session['amount']
                usage['sessions'].append({
                    'id': index,
                    'account_addr': str(session['from_addr']).lower(),
                    'received_bytes': session['sent_bytes'],
                    'session_duration': session['session_duration'],
                    'amount': session['amount'],
                    'timestamp': session['timestamp']
                })

            return error, usage

        error, sessions_count = self.get_vpn_sessions_count(account_addr)
        if error is None:
            for index in range(0, sessions_count):
                session_id = get_encoded_session_id(account_addr, index)
                error, _usage = vpn_service_manager.get_vpn_usage(account_addr, session_id)
                if error is None:
                    if _usage[5] is False:
                        usage['due'] += _usage[3]
                    usage['stats']['received_bytes'] += _usage[1]
                    usage['stats']['duration'] += _usage[2]
                    usage['stats']['amount'] += _usage[3]
                    usage['sessions'].append({
                        'id': session_id,
                        'account_addr': str(_usage[0]).lower(),
                        'received_bytes': _usage[1],
                        'session_duration': _usage[2],
                        'amount': _usage[3],
                        'timestamp': _usage[4],
                        'is_paid': _usage[5]
                    })

        return error, usage

    def pay_vpn_session(self, from_addr, session_id, net, tx_data, payment_type, device_id=None):
        errors, tx_hashes = [], []
        error, tx_hash = self.raw_transaction(tx_data, net)
        if error is None:
            tx_hashes.append(tx_hash)
            if payment_type == 'init':
                error, tx_hash = vpn_service_manager.set_initial_payment(from_addr)
            elif payment_type == 'normal':
                error, _usage = vpn_service_manager.get_vpn_usage(from_addr, session_id)
                error, tx_hash = vpn_service_manager.pay_vpn_session(from_addr, int(_usage[3]), session_id)
            if error is None:
                tx_hashes.append(tx_hash)
            else:
                errors.append(error)
        else:
            errors.append(error)

        return errors, tx_hashes

    def add_vpn_usage(self, from_addr, to_addr, sent_bytes, session_duration, amount, timestamp, device_id=None):
        _sent_bytes, _session_duration, _amount = sent_bytes, session_duration, amount
        error, tx_hash, make_tx, session_id = None, None, False, None
        error, sessions_count = self.get_vpn_sessions_count(to_addr)  # to_addr: client account address
        if error is None:
            session_id = get_encoded_session_id(to_addr, sessions_count)
            _usage = db.usage.find_one({
                'from_addr': from_addr,
                'to_addr': to_addr
            })
            if _usage is None:
                if (sent_bytes > LIMIT_10MB) and (sent_bytes < LIMIT_100MB):
                    _ = db.usage.insert_one({
                        'from_addr': from_addr,
                        'to_addr': to_addr,
                        'sent_bytes': sent_bytes,
                        'session_duration': session_duration,
                        'amount': amount,
                        'timestamp': timestamp
                    })
                elif sent_bytes >= LIMIT_100MB:
                    make_tx = True
            elif (_usage['sent_bytes'] + sent_bytes) < LIMIT_100MB:
                _ = db.usage.find_one_and_update({
                    'from_addr': from_addr,
                    'to_addr': to_addr
                }, {
                    '$set': {
                        'sent_bytes': _usage['sent_bytes'] + sent_bytes,
                        'session_duration': _usage['session_duration'] + session_duration,
                        'amount': _usage['amount'] + amount,
                        'timestamp': timestamp
                    }
                })
            else:
                sent_bytes = int(_usage['sent_bytes'] + sent_bytes)
                session_duration = int(
                    _usage['session_duration'] + session_duration)
                amount = int(_usage['amount'] + amount)
                make_tx = True
                _ = db.usage.find_one_and_delete({
                    'from_addr': from_addr,
                    'to_addr': to_addr
                })

        if device_id and to_addr == REFERRAL_DUMMY:
            _ = db.ref_sessions.insert_one({
                'device_id': device_id,
                'session_id': session_id,
                'from_addr': from_addr,
                'to_addr': to_addr,
                'sent_bytes': _sent_bytes,
                'session_duration': _session_duration,
                'amount': _amount,
                'timestamp': timestamp
            })
            if _sent_bytes >= LIMIT_100MB:
                _, res = add_bonus(device_id)

        if make_tx is True and to_addr != REFERRAL_DUMMY:
            error, tx_hash = vpn_service_manager.add_vpn_usage(from_addr, to_addr, sent_bytes, session_duration, amount,
                                                               timestamp, session_id)

        return error, tx_hash

    # DEV
    def free(self, to_addr, eths, sents):
        errors, tx_hashes = [], []
        error, tx_hash = self.transfer_eths(COINBASE_ADDRESS, to_addr, eths, COINBASE_PRIVATE_KEY, 'rinkeby')

        if error is None:
            tx_hashes.append(tx_hash)
            error, tx_hash = self.transfer_erc20(COINBASE_ADDRESS, to_addr, sents, 'SENT', COINBASE_PRIVATE_KEY,
                                                 'rinkeby')
            if error is None:
                tx_hashes.append(tx_hash)
            else:
                errors.append(error)
        else:
            errors.append(error)

        return errors, tx_hashes


eth_helper = ETHHelper()

File: master-node-nodejs/src/helpers/eth.js
import md5 from 'md5'
import redis from 'redis'
import async from 'async'

import {
  addSession
} from "./referral";
import {
  Eth_manager
} from '../eth/eth';
import {
  ERC20Manager
} from '../eth/erc20';
import {
  VpnServiceManager
} from '../eth/vpn_contract';
import {
  LIMIT_10MB,
  LIMIT_100MB,
  SESSIONS_SALT
} from '../config/vars';
import {
  ADDRESS as COINBASE_ADDRESS,
  PRIVATE_KEY as COINBASE_PRIVATE_KEY
} from '../config/eth';
import {
  Usage,
  refSession
} from '../models';
import database from '../db/database';
import {
  REFERRAL_DUMMY
} from '../config/referral';

let redisClient = redis.createClient();

const getEncodedSessionId = async (accountAddr, index, cb) => {
  accountAddr = accountAddr.toString('utf8');
  index = index.toString();
  index = index.toString('utf8')
  let sessionId = await md5(accountAddr + index + SESSIONS_SALT)
  return cb(sessionId)
}

const createAccount = (password, cb) => {
  Eth_manager['main'].createAccount(password, (err, accountDetails) => {
    cb(err, accountDetails);
  });
}

const getAccountAddress = (privateKey, cb) => {
  Eth_manager['main'].getAddress(privateKey,
    (err, address) => {
      let accountAddress = address.substr(2)
      cb(null, accountAddress)
    })
}

const getTxReceipt = (txHash, net, cb) => {
  if (net === 'main') {
    Eth_manager['main'].getTransactionReceipt(txHash, (err, receipt) => {
      cb(err, receipt)
    })
  } else if (net === 'rinkeby') {
    Eth_manager['rinkeby'].getTransactionReceipt(txHash, (err, receipt) => {
      cb(err, receipt)
    })
  }
}

const getTx = (txHash, net, cb) => {
  if (net === 'main') {
    Eth_manager['main'].getTransaction(txHash, (err, receipt) => {
      cb(err, receipt)
    })
  } else if (net === 'rinkeby') {
    Eth_manager['rinkeby'].getTransaction(txHash, (err, receipt) => {
      cb(err, receipt)
    })
  }
}

const getTxCount = (accountAddr, net, cb) => {
  if (net == 'main') {
    Eth_manager['main'].getTransactionCount(accountAddr, (err, txCount) => {
      cb(txCount)
    })
  } else if (net == 'rinkeby') {
    Eth_manager['rinkeby'].getTransactionCount(accountAddr, (err, txCount) => {
      cb(txCount)
    })
  }
}

const getValidNonce = (accountAddr, net, cb) => {
  let key = accountAddr + '_' + net;
  let error = null;
  let nonce = 0;

  redisClient.get(key, (error, previousNonce) => {
    if (previousNonce)
      previousNonce = parseInt(previousNonce);

    if (net == 'main') {
      Eth_manager['main'].getTransactionCount(accountAddr, (err, nonce) => {
        if ((err === null) && (previousNonce === null || nonce > previousNonce)) {
          redisClient.set(key, nonce)
          cb(nonce)
        } else {
          setTimeout(() => {
            getValidNonce(accountAddr, net, cb)
          }, 1000);
        }
      })
    } else if (net == 'rinkeby') {
      Eth_manager['rinkeby'].getTransactionCount(accountAddr, (err, nonce) => {
        if ((err === null) && (previousNonce === null || nonce > previousNonce)) {
          redisClient.set(key, nonce)
          cb(nonce)
        } else {
          setTimeout(() => {
            getValidNonce(accountAddr, net, cb)
          }, 1000);
        }
      })
    }
  });
}

const getBalances = (accountAddr, cb) => {
  let balances = {
    main: {
      eths: null,
      sents: null
    },
    test: {
      eths: null,
      sents: null
    }
  }

  try {
    Eth_manager['main'].getBalance(accountAddr, (err, balance) => {
      balances.main.eths = balance
      ERC20Manager['main']['SENT'].getBalance(accountAddr, (err, balance) => {
        balances.main.sents = balance
        Eth_manager['rinkeby'].getBalance(accountAddr, (err, balance) => {
          balances.test.eths = balance
          ERC20Manager['rinkeby']['SENT'].getBalance(accountAddr, (err, balance) => {
            balances.test.sents = balance
            cb(null, balances);
          })
        })
      })
    })
  } catch (error) {
    cb(error, null)
  }
}

const transfer = (fromAddr, toAddr, amount, symbol, privateKey, net, cb) => {
  if (symbol == 'ETH') {
    transferEths(fromAddr, toAddr, amount, privateKey, net,
      (err, resp) => {
        cb(err, resp)
      })
  } else {
    transferErc20(fromAddr, toAddr, amount, symbol, privateKey, net,
      (err, resp) => {
        cb(err, resp);
      })
  }
}

const transferErc20 = (fromAddr, toAddr, amount, symbol, privateKey, net, cb) => {
  getValidNonce(fromAddr, net, (nonce) => {
    ERC20Manager[net][symbol].transferAmount(toAddr, amount, privateKey, nonce, (err, txHash) => {
      cb(err, txHash)
    })
  })
}

const transferEths = (fromAddr, toAddr, amount, privateKey, net, cb) => {
  if (net == 'main') {
    Eth_manager['main'].transferAmount(fromAddr, toAddr, amount, privateKey, (err, txHash) => {
      cb(err, txHash)
    })
  } else if (net == 'rinkeby') {
    Eth_manager['rinkeby'].transferAmount(fromAddr, toAddr, amount, privateKey, (err, txHash) => {
      cb(err, txHash)
    })
  }
}

const rawTransaction = (txData, net, cb) => {
  if (net == 'main') {
    Eth_manager['main'].sendRawTransaction(txData,
      (err, txHash) => {
        cb(err, txHash);
      })
  } else if (net == 'rinkeby') {
    Eth_manager['rinkeby'].sendRawTransaction(txData,
      (err, txHash) => {
        cb(err, txHash);
      })
  }
}

const getInitialPayment = (accountAddr, cb) => {
  VpnServiceManager.getInitialPayment(accountAddr, (err, isPayed) => {
    cb(err, isPayed)
  })
}

const getDueAmount = (accountAddr, cb) => {
  VpnServiceManager.getDueAmount(accountAddr,
    (err, dueAmount) => {
      cb(err, dueAmount);
    });
}

const getVpnSessionCount = (accountAddr, cb) => {
  VpnServiceManager.getVpnSessionCount(accountAddr, (err, sessions) => {
    cb(err, sessions);
  })
}

const getLatestVpnUsage = (accountAddr, cb) => {
  getVpnSessionCount(accountAddr, (err, sessionsCount) => {
    if (!err && sessionsCount > 0) {
      getEncodedSessionId(accountAddr, sessionsCount - 1, (sessionId) => {
        VpnServiceManager.getVpnUsage(accountAddr, sessionId, (err, _usage) => {
          if (!err) {
            let usage = {
              'id': sessionId,
              'account_addr': _usage[0].toString().toLowerCase(),
              'received_bytes': _usage[1],
              'session_duration': _usage[2],
              'amount': _usage[3],
              'timestamp': _usage[4],
              'is_paid': _usage[5]
            }
            cb(null, usage)
          } else {
            cb(err, null)
          }
        })
      })
    }
  })
}

const getVpnUsage = async (accountAddr, cb) => {
  let usage = {
    'due': 0,
    'stats': {
      'received_bytes': 0,
      'duration': 0,
      'amount': 0
    },
    'sessions': []
  }
  VpnServiceManager.getVpnSessionCount(accountAddr, (err, sessions) => {
    if (!err) {
      async.times(sessions, (index, next) => {
        getEncodedSessionId(accountAddr, index, (sessionId) => {
          VpnServiceManager.getVpnUsage(accountAddr, sessionId, (error, _usage) => {
            if (!error) {
              if (!_usage[5])
                usage['due'] += _usage[3]
              usage['stats']['received_bytes'] += parseInt(_usage[1])
              usage['stats']['duration'] += parseInt(_usage[2])
              usage['stats']['amount'] += parseInt(_usage[3])
              usage['sessions'].push({
                'id': sessionId,
                'account_addr': _usage[0],
                'received_bytes': _usage[1],
                'session_duration': _usage[2],
                'amount': _usage[3],
                'timestamp': _usage[4],
                'is_paid': _usage[5]
              })
              next()
            } else {
              next()
              // cb(error, null)
            }
          })
        })
      }, () => {
        cb(null, usage)
      })
    } else {
      cb(err, null)
    }
  })
}

const payVpnSession = (fromAddr, amount, sessionId, net, txData, paymentType, deviceId = null, cb) => {
  let errors = []
  let txHashes = []

  rawTransaction(txData, net, (err1, txHash1) => {
    if (!err1) {
      txHashes.push(txHash1)
      getValidNonce(COINBASE_ADDRESS, 'rinkeby', (nonce) => {
        if (paymentType == 'init') {
          VpnServiceManager.setInitialPayment(fromAddr, nonce, (err2, txHash2) => {
            if (!err2) {
              if (deviceId) {
                addSession(deviceId, sessionId, txHash2, (_, res) => {
                  txHashes.push(txHash2)
                  cb(errors, txHashes)
                })
              }
            } else {
              errors.push(err2)
              cb(errors, txHashes)
            }
          })
        } else if (paymentType == 'normal') {
          VpnServiceManager.payVpnSession(fromAddr, amount, sessionId, nonce, (err2, txHash2) => {
            if (!err2) {
              if (deviceId) {
                addSession(deviceId, sessionId, txHash2, (_, res) => {
                  txHashes.push(txHash2)
                  cb(errors, txHashes)
                })
              }
              txHashes.push(txHash2)
              cb(errors, txHashes)
            } else {
              errors.push(err2)
              cb(errors, txHashes)
            }
          })
        } else {
          cb(errors, txHashes)
        }
      })
    } else {
      errors.push(err1)
      cb(errors, txHashes);
    }
  })
}

const addVpnUsage = (fromAddr, toAddr, sentBytes, sessionDuration, amount, timeStamp, deviceId, cb) => {
  let err = null;
  let txHash = null;
  let makeTx = false;
  let sessionId = null;
  let _usage = null;

  sentBytes = parseInt(sentBytes);
  sessionDuration = parseInt(sessionDuration);
  amount = parseInt(amount);

  async.waterfall([
    (next) => {
      getVpnSessionCount(toAddr, (err, sessionsCount) => {
        if (!err) {
          getEncodedSessionId(toAddr, sessionsCount, (sessId) => {
            sessionId = sessId;
            next();
          })
        } else {
          next();
        }
      })
    }, (next) => {
      Usage.findOne({
        'from_addr': fromAddr,
        'to_addr': toAddr
      }, (err, usage) => {
        _usage = usage
        next()
      })
    }, (next) => {
      if (!_usage) {
        if (sentBytes > LIMIT_10MB && sentBytes < LIMIT_100MB) {

          let data = {
            'from_addr': fromAddr,
            'to_addr': toAddr,
            'sent_bytes': sentBytes,
            'session_duration': sessionDuration,
            'amount': amount,
            'timestamp': timeStamp
          }

          let usageData = new Usage(data)

          database.insert(usageData, (err, resp) => {
            if (err) {
              next(err, null)
            } else {
              next()
            }
          })
        } else if (sentBytes >= LIMIT_100MB) {
          makeTx = true;
          next()
        } else {
          next()
        }
      } else if (parseInt(_usage['sent_bytes']) + sentBytes < LIMIT_100MB) {

        let findData = {
          'from_addr': fromAddr,
          'to_addr': toAddr
        }

        let updateData = {
          'sent_bytes': _usage['sent_bytes'] + sentBytes,
          'session_duration': _usage['session_duration'] + sessionDuration,
          'amount': _usage['amount'] + amount,
          'timestamp': timeStamp
        }

        database.update(Usage, findData, updateData, (err, resp) => {
          next()
        })
      } else {
        sentBytes = parseInt(_usage['sent_bytes']) + sentBytes;
        sessionDuration = parseInt(_usage['session_duration']) + sessionDuration;
        amount = parseInt(_usage['amount']) + amount;
        makeTx = true;
        Usage.findOneAndRemove({
          'from_addr': fromAddr,
          'to_addr': toAddr
        }, (err, resp) => {
          next()
        })
      }
    }, (next) => {
      if (toAddr === REFERRAL_DUMMY && sentBytes >= LIMIT_100MB) {
        addSession(deviceId, sessionId, null, (_, resp) => {
          let error = _
          let _refSession = new refSession({
            'device_id': deviceId,
            'session_id': deviceId,
            'from_addr': fromAddr,
            'to_addr': toAddr,
            'sent_bytes': sentBytes,
            'session_duration': sessionDuration,
            'amount': amount,
            'timestamp': timeStamp
          })
          _refSession.save((err, resp) => {
            next(err, resp)
          })
        })
      } else {
        next(null)
      }
      if (makeTx && toAddr !== REFERRAL_DUMMY) {
        getValidNonce(COINBASE_ADDRESS, 'rinkeby', (nonce) => {
          VpnServiceManager.addVpnUsage(fromAddr, toAddr, sentBytes, sessionDuration, amount, timeStamp, sessionId, nonce,
            (err, txHash) => {
              next(err, txHash)
            })
        })
      } else {
        next(null, null)
      }
    }
  ], (err, resp) => {
    cb(err, resp);
  })
}

const free = (toAddr, eths, sents, cb) => {
  let errors = [],
    txHashes = []

  let PRIVATE_KEY = Buffer.from(COINBASE_PRIVATE_KEY, 'hex');
  transferEths(COINBASE_ADDRESS, toAddr, eths, PRIVATE_KEY, 'rinkeby', (err, txHash) => {
    if (!err) {
      txHashes.push(txHash);
      transferErc20(COINBASE_ADDRESS, toAddr, sents, 'SENT', PRIVATE_KEY, 'rinkeby', (err, txHash) => {
        if (!err) {
          txHashes.push(txHash);
          cb(errors, txHashes);
        } else {
          errors.push(errors);
          cb(errors, txHashes);
        }
      })
    } else {
      errors.push(err);
      cb(errors, txHashes);
    }
  });
}

export default {
  createAccount,
  getAccountAddress,
  getTxReceipt,
  getTx,
  getTxCount,
  getValidNonce,
  getBalances,
  transfer,
  transferErc20,
  transferEths,
  rawTransaction,
  getInitialPayment,
  getDueAmount,
  getVpnSessionCount,
  getLatestVpnUsage,
  getVpnUsage,
  payVpnSession,
  addVpnUsage,
  free,
}
